optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/chained/TailChainSwapMove.java

```diff
@@ -38,20 +38,124 @@
 public class TailChainSwapMove<Solution_> extends AbstractMove<Solution_> {
 
     protected final GenuineVariableDescriptor<Solution_> variableDescriptor;
-    protected final SingletonInverseVariableSupply inverseVariableSupply;
-    protected final AnchorVariableSupply anchorVariableSupply;
 
     protected final Object leftEntity;
+    protected final Object leftValue;
+    protected final Object leftAnchor;
+
+    protected final Object rightEntity; // Sometimes null
     protected final Object rightValue;
+    protected final Object rightAnchor;
+
+    protected final boolean sameAnchor;
+    protected final Object leftNextEntity;
+    protected final Object rightNextEntity;
+    protected final boolean reverseAnchorSide;
+    protected final Object lastEntityInChain;
+    protected final Object entityAfterAnchor;
 
     public TailChainSwapMove(GenuineVariableDescriptor<Solution_> variableDescriptor,
             SingletonInverseVariableSupply inverseVariableSupply, AnchorVariableSupply anchorVariableSupply,
             Object leftEntity, Object rightValue) {
         this.variableDescriptor = variableDescriptor;
-        this.inverseVariableSupply = inverseVariableSupply;
-        this.anchorVariableSupply = anchorVariableSupply;
         this.leftEntity = leftEntity;
+        leftValue = variableDescriptor.getValue(leftEntity);
+        leftAnchor = anchorVariableSupply.getAnchor(leftEntity);
+        rightEntity = inverseVariableSupply.getInverseSingleton(rightValue);
+        this.rightValue = rightValue;
+        rightAnchor = variableDescriptor.isValuePotentialAnchor(rightValue) ? rightValue
+                : anchorVariableSupply.getAnchor(rightValue);
+        sameAnchor = leftAnchor == rightAnchor;
+        if (!sameAnchor) {
```

Comment:
It's more natural to have 'positive path' in the `if` branch and 'negative' in the `else` branch. Basically, if there is an if-else statement, you shouldn't use negation in the condition.